File: backend/app/routers/recommendations.py
from fastapi import APIRouter, HTTPException

from app.core.logging import get_logger
from app.models.schemas import RecommendationRequest, RecommendationResponse
from app.services.recommendations import fetch_next_purchase

router = APIRouter(tags=["recommendations"])
logger = get_logger(__name__)


@router.post("/recommendations/next-purchase", response_model=RecommendationResponse)
async def next_purchase(req: RecommendationRequest) -> RecommendationResponse:
    user_input = req.user_input.strip() or "Give recommendations"
    try:
        return await fetch_next_purchase(user_input)
    except Exception as exc:
        logger.exception("Recommendation endpoint error")
        raise HTTPException(
            status_code=502,
            detail="The recommendation service is currently unavailable. Please try again.",
        ) from exc

File: worker/src/routers/recommendations.ts
import { Hono } from "hono";

import type { Env } from "../env";
import { loadSettings } from "../lib/config";
import { fetchNextPurchase } from "../lib/recommendations";
import { getSalesforceClient } from "../lib/salesforce";
import type { RecommendationRequest } from "../models/schemas";

export const recommendationsRouter = new Hono<{ Bindings: Env }>();

recommendationsRouter.post("/api/recommendations/next-purchase", async (c) => {
  const req = await c.req.json<RecommendationRequest>();
  const s = loadSettings(c.env);
  const userInput = (req.user_input ?? "").trim() || "Give recommendations";
  try {
    return c.json(await fetchNextPurchase(s, getSalesforceClient(s), userInput, req.refresh ?? false));
  } catch {
    return c.json(
      { detail: "The recommendation service is currently unavailable. Please try again." },
      502,
    );
  }
});
